service/history/commandChecker.go

```diff
@@ -173,6 +172,51 @@ func (c *workflowSizeChecker) checkIfMemoSizeExceedsLimit(
 	return nil
 }
 
+func withinLimit(value int, limit int) bool {
+	if limit <= 0 {
+		// limit not defined
+		return true
+	}
+	return value < limit
+}
+
+func (c *workflowSizeChecker) checkIfNumChildWorkflowsExceedsLimit() error {
+	key := c.mutableState.GetWorkflowKey()
+	logger := log.With(
+		c.logger,
+		tag.WorkflowNamespaceID(key.NamespaceID),
+		tag.WorkflowID(key.WorkflowID),
+		tag.WorkflowRunID(key.RunID),
+	)
+
+	numPending := len(c.mutableState.GetPendingChildExecutionInfos())
+	errLimit := c.numPendingChildExecutionLimitError
+	warnLimit := c.numPendingChildExecutionLimitWarn
+	if !withinLimit(numPending, errLimit) {
+		c.metricsHandler.Counter(metrics.NumPendingChildWorkflowsTooHigh.GetMetricName()).Record(1)
+		err := fmt.Errorf(
+			"the number of pending child workflow executions, %d, "+
+				"has reached the error limit of %d established with %q",
```

Comment:
Nit: I think we should give a fixed error message and this can be error log message. And add the namespace id, workflow id run id tags so that we can identify those workflows if it happens.

Reply:
I think given clear limit count and current pending count is very useful. The log should include namespace / workflow_id / run_id.

Reply:
nit: Ideally, log message should be static and use tags for other information like limit, etc. It will make it much easier why filtering logs by msg.

Reply:
Is my understanding correct that `numPending` will always equal to `errLimit` as child workflows are added to mutable state one by one.